docs/proposals/20210628-workloadspread.md

```diff
@@ -0,0 +1,287 @@
+
+Controller will mark the subset containing unscheduable pods to unscheduable status and Webhook cannot inject Pod into this subset by check subset's status.
+And then controller cleans up all unscheduable Pods to trigger workload creating new replicas, and webhook will skip unscheduable subset and chose suitable subset to inject. 
+
+The unscheduled subset can be kept for 10 minutes and then should be recovered schedulable to schedule Pod again by controller.
+
+#### Deletion priority
+
+We have three types for subset's Pod deletion-cost
+1. the number of active Pods in this subset <= maxReplicas, deletion-cost = 100. indicating the priority of Pods \
+   in this subset is higher than other Pods in workload.
+2. subset's maxReplicas is nil, deletion-cost = 0. indicating the priority of Pods in this subset is lower than \
+   other subsets that have a not nil maxReplicas.
+3. the number of active Pods in this subset > maxReplicas, two class: (a) deletion-cost = -100, (b) deletion-cost = +100. \
+   indicating we prefer deleting the Pod have -100 deletion-cost in this subset in order to control the instance of subset \
+   meeting up the desired maxReplicas number.
+
+CloneSet has supported deletion-cost feature in the recent versions.
+
+Since Kubernetes 1.21, there is a new annotation definition on Pod:
+
+```yaml
+    controller.kubernetes.io/pod-deletion-cost: -100 / 100 / 0
+```
+
+(In 1.21, users need to enable `PodDeletionCost` feature-gate, and since 1.22 it will be enabled by default)
+
+### Alternative Considered
+
+#### UnitedDeployment
+
+Both UnitedDeployment and WorkloadSpread allow the workload to be distributed in different subsets. 
+The topology distribution of UnitedDeployment is defined in advance by the user, but the topology distribution of WorkloadSpread has a sequential attribute. 
+Only if the number of Pods of the current Subset is equal to maxReplicas, the remaining replicas will be scheduled to the next Subset. 
+The order attribution of subset makes WorkloadSpread better meeting up the elastic demand of workload. 
+
+For example, WorkloadSpread has two subsets: subset-a and subset-b, maxReplicas are 50 and null(no limit). 
+The number of replicas of workload will change with the actual business load. 
+
+workload replicas changes: 40 -> 100 -> 30
+
+subset replicas distribution:
+
+|subset-a(50) |subset-b(no limit)|
+|----| ----|
+| 40  | 0  |
+| 50  | 50 |
+| 30  | 0  |
+
+The subset-b will only be allocated Pods when the application load is higher (100 replicas), and not provide resources during the rest of the time.
+This will reduce a lot of resources belongs to subset-b for us, and the application can get better elastic support.
+
+#### NodeAffinity
+
+Add nodeAffinity to the workload template, such as preferredDuringSchedulingIgnoredDuringExecution, which can be scheduled in multiple regions.
+However, It cannot specify the replica numbers of the subset, and it is only valid when the Pod is scheduled.
```

Comment:
>  It cannot specify the replica numbers of the subset, and it is only valid when the Pod is scheduled.
it cannot limit the replica numbers of a subset, and it is only effective when scaling out the workload, not effective when scaling in